### PR TITLE
Tls/auto fallback

### PR DESCRIPTION
**Pull request by nejisama**
1. fix tls timeout causes deadloop
2. add client tls fallback when handshake failed.

### CODE REVIEW

File: pkg/mtls/conn.go [hunk 1]
```diff
@@ -41,8 +41,14 @@ type TLSConn struct {
...
 	n, err := c.Conn.Read(b)
-	if err != nil && strings.Contains(err.Error(), "tls") {
-		log.DefaultLogger.Alertf(types.ErrorKeyTLSRead, "[mtls] tls connection read error: %v, local address: %v, remote address: %v", err, c.Conn.LocalAddr(), c.Conn.RemoteAddr())
+	if err != nil {
+		if strings.Contains(err.Error(), "tls") {
+			log.DefaultLogger.Alertf(types.ErrorKeyTLSRead, "[mtls] tls connection read error: %v, local address: %v, remote address: %v", err, c.Conn.LocalAddr(), c.Conn.RemoteAddr())
+		}
+		if !c.Conn.GetConnectionState().HandshakeComplete {
+			// wraps as a new error which makes the read error and do no retry
+			return n, errors.New("tls: handshake is not completed: %s" + err.Error())
```
Comment (taoyuanyuan) [inline pkg/mtls/conn.go:50]: delete %s

File: pkg/network/connection.go [hunk 2]
```diff
@@ -1067,78 +1067,80 @@ func NewClientConnection(connectTimeout time.Duration, tlsMng types.TLSContextMa
...
-			}
+func (cc *clientConnection) tryConnect() (event api.ConnectionEvent, err error) {
+	event, err = cc.connect()
+	if err != nil {
+		return event, err
+	}
+	if cc.tlsMng == nil {
+		return event, err
+	}
+	cc.rawConnection, err = cc.tlsMng.Conn(cc.rawConnection)
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:1125]: maybe panic when tlsMng.Conn return (nil, error).

**Review by taoyuanyuan — APPROVED** [review]
LGTM

File: pkg/network/connection.go [hunk 3]
```diff
@@ -1067,78 +1067,80 @@ func NewClientConnection(connectTimeout time.Duration, tlsMng types.TLSContextMa
...
+	if err == nil {
```
Comment (wangfakang) [inline pkg/network/connection.go:1126]: it should be  `return event, nil`.

File: pkg/network/connection.go [hunk 4]
```diff
@@ -1067,78 +1067,80 @@ func NewClientConnection(connectTimeout time.Duration, tlsMng types.TLSContextMa
...
+}
 
-			}
+func (cc *clientConnection) tryConnect() (event api.ConnectionEvent, err error) {
+	event, err = cc.connect()
+	if err != nil {
+		return event, err
+	}
+	if cc.tlsMng == nil {
+		return event, err
```
Comment (wangfakang) [inline pkg/network/connection.go:1123]: it should be `return event, nil`.

File: pkg/network/connection.go [hunk 5]
```diff
@@ -1067,78 +1067,80 @@ func NewClientConnection(connectTimeout time.Duration, tlsMng types.TLSContextMa
...
+	switch cc.network {
+	case "udp":
+		if tc, ok := cc.rawConnection.(*net.UDPConn); ok {
+			cc.file, err = tc.File()
+		}
+	case "unix":
+		if tc, ok := cc.rawConnection.(*net.UnixConn); ok {
+			cc.file, err = tc.File()
+		}
+	case "tcp":
```
Comment (wangfakang) [inline pkg/network/connection.go:1109]: It should be call `c.cc.rawConnection.Close()` when use `UseNetpollMode`.

Comment (nejisama) [reply pkg/network/connection.go:1109]: no logic changed.

**Review by wangfakang — APPROVED** [review]

**Pull request merged**